src/scikit_build_core/resources/_editable_redirect.py

```diff
@@ -1,59 +1,340 @@
 
 DIR = os.path.abspath(os.path.dirname(__file__))
 MARKER = "SKBUILD_EDITABLE_SKIP"
 VERBOSE = "SKBUILD_EDITABLE_VERBOSE"
 
 __all__ = ["install"]
 
+# TODO: Probably not the correct syntax?
+SKBuildTree = Dict[str, List[Traversable]]
+
 
 def __dir__() -> list[str]:
     return __all__
 
 
-class ScikitBuildRedirectingFinder(importlib.abc.MetaPathFinder):
+# Loaders:
+# - These handle the loading of the python source or library extension
+# - File navigation is done through the reader in get_resource_reader
+# - Loaders and reader creation are relative to the module path, not the files themselves
+
+
+class SKBuildLoader:
+    """
+    Base loader with the common implementation of the navigation
+    """
+
+    # TODO: find the base class to subclass this from for `ModuleSpec` constructor
+    def __init__(self, tree: SKBuildTree):
+        self._tree = tree
+
+    def get_resource_reader(self, module: str) -> TraversableResources:
+        return SKBuildReader(self._tree, module)
+
+    # Notes: See implementation of spec_from_file_location.
+    # default loader should not implement `get_filename` or `is_package`
+    # submodule_search_locations will be set to empty list by default
+
+
+class SKBuildSourceLoader(SourceFileLoader, SKBuildLoader):
+    """
+    Loads Python source files (.py)
+    """
+
+    def __init__(self, tree: SKBuildTree, name: str, path: str):
+        SKBuildLoader.__init__(self, tree)
+        SourceFileLoader.__init__(self, name, path)
+
+    def set_data(self, *args: Any, **kwargs: Any) -> None:
+        # TODO: Have to set to pass. Is it ideal?
+        pass
+
+    def get_resource_reader(self, module: str) -> TraversableResources:
+        return SKBuildLoader.get_resource_reader(self, module)
+
+
+class SKBuildSourcelessLoader(SourcelessFileLoader, SKBuildLoader):
+    def __init__(self, tree: SKBuildTree, name: str, path: str):
+        SKBuildLoader.__init__(self, tree)
+        SourcelessFileLoader.__init__(self, name, path)
+
+    def get_resource_reader(self, module: str) -> TraversableResources:
+        return SKBuildLoader.get_resource_reader(self, module)
+
+
+class SKBuildExtensionLoader(ExtensionFileLoader, SKBuildLoader):
+    def __init__(self, tree: SKBuildTree, name: str, path: str):
+        SKBuildLoader.__init__(self, tree)
+        ExtensionFileLoader.__init__(self, name, path)
+
+    def get_resource_reader(self, module: str) -> TraversableResources:
+        # TODO: Investigate extension's TraversableResources?
+        return SKBuildLoader.get_resource_reader(self, module)
+
+
+class SKBuildNamespaceLoader(ExecutionLoader, SKBuildLoader):
+    def get_source(self, fullname: str) -> str:
+        return ""
+
+    def get_filename(self, fullname: str) -> str:
+        return ""
+
+    def __init__(self, tree: SKBuildTree):
+        SKBuildLoader.__init__(self, tree)
+
+    def exec_module(self, module: Any) -> None:
+        pass
+
+    def create_module(self, spec: Any) -> Any:
```

Comment:
Oh, if we return None, a module is still created, but without the appropriate metadata. I think this is where we need to edit __path__ right?

Reply:
Ooh, `__path__` is an iterable. So in namespace package it should be all of the filepaths where it is aliasing. And if my intuition is correct, submodule_search_path is equivalent. If I understand that correctly, we can simply inject `__path__`s for the packages/namespaces pointing to the source and the build/install directory and the system will pick up the modules in there.

Granted it might not work for `importlib.resources` 🤔

Reply:
Yes, `__path__` is an iterable that describes all of the locations that the module is spread over. It's used for namespace packages which are allowed to be spread across different directories. Though I think `__path__` is mostly informational and optional.

Reply:
(No need to worry too much about namespace package too much yet, just fixing importlib.resources on normal modules is a huge step.)